In [2]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [3]:
accident = pd.read_csv('datasets\\uk_road_accident.csv')

In [4]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h2>Descriptive Analytics</h2>

In [5]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

<h2>Checking Null Values</h2>

In [6]:
accident.isnull()

Index  Accident_Severity  Accident Date  Latitude  Light_Conditions  \
0       False              False          False     False             False   
1       False              False          False     False             False   
2       False              False          False     False             False   
3       False              False          False     False             False   
4       False              False          False     False             False   
...       ...                ...            ...       ...               ...   
660674  False              False          False     False             False   
660675  False              False          False     False             False   
660676  False              False          False     False             False   
660677  False              False          False     False             False   
660678  False              False          False     False             False   

        District Area  Longitude  Number_of_Casualties  Number_of_Vehicles  \
0               False      False                 False               False   
1               False      False                 False               False   
2               False      False                 False               False   
3               False      False                 False               False   
4               False      False                 False               False   
...               ...        ...                   ...                 ...   
660674          False      False                 False               False   
660675          False      False                 False               False   
660676          False      False                 False               False   
660677          False      False                 False               False   
660678          False      False                 False               False   

        Road_Surface_Conditions  Road_Type  Urban_or_Rural_Area  \
0                         False      False                False   
1                         False      False                False   
2                         False       True                False   
3                         False      False                False   
4                         False       True                False   
...                         ...        ...                  ...   
660674                    False      False                False   
660675                    False      False                False   
660676                    False      False                False   
660677                    False      False                False   
660678                    False      False                False   

        Weather_Conditions  Vehicle_Type  
0                    False         False  
1                    False         False  
2                     True         False  
3                    False         False  
4                     True         False  
...                    ...           ...  
660674               False         False  
660675               False         False  
660676               False         False  
660677               False         False  
660678               False         False  

[660679 rows x 14 columns]

In [7]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [8]:
accident['Accident_Severity']

0         Serious
1         Serious
2         Serious
3         Serious
4          Slight
           ...   
660674     Slight
660675     Slight
660676     Slight
660677    Serious
660678    Serious
Name: Accident_Severity, Length: 660679, dtype: object

In [9]:
accident['Accident_Severity'].unique()

array(['Serious', 'Slight', 'Fatal'], dtype=object)

<h1>QUESTIONS</h1>

<h2>1.How many accidents are there under each severity?</h2>

In [10]:
accident['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h2>INSIGHT:</h2>


<h3>There are 563,801 accidents under slight severity, serious with 88,217 and fatal with 8,661.</h3>

<h2>2.Which district area has the most accident report?</h2>

In [11]:
accident['District Area'].value_counts()

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h2>INSIGHT:</h2>

<h3>Based on the data ,the district area that has the most accident report is Birmingham with a total of 1349. </h3>

<h2>3.What is the average number of casualties fall under serious accident severity?</h2>

In [12]:
accident[accident['Accident_Severity'] =='Serious']['Number_of_Casualties'].mean()

np.float64(1.46727954929322)

<h2>INSIGHT:</h2>

<h3>The result shows that serious accidents usually affect one person with the average result of 1.47 it means that more than one casualty on average was involved.</h3>

<h2>4.Where are road accidents higher, in urban or rural areas?</h2>

In [13]:
accident['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421663
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h2>Insight:</h2>

<h3>Based on the result road accidents are much higher in Urban areas with 421663 compared to rural areas with 238990.</h3>

<h2>5.How many slight accidents occured on frost or ice compared to snow?</h2>

In [14]:
accident.groupby(['Road_Surface_Conditions','Accident_Severity']).size()

Road_Surface_Conditions  Accident_Severity
Dry                      Fatal                  5788
                         Serious               61638
                         Slight               380395
Flood over 3cm. deep     Fatal                    23
                         Serious                 152
                         Slight                  842
Frost or ice             Fatal                   193
                         Serious                2007
                         Slight                16317
Snow                     Fatal                    35
                         Serious                 565
                         Slight                 5290
Wet or damp              Fatal                  2620
                         Serious               23785
                         Slight               160303
dtype: int64

<h2>INSIGHT:</h2>

<h3>Slight accident on frost or ice shows a large number of accidents with a total of 16317 compared to snow with 5290 accidents. </h3>

<h2>6.Which specific date recorded the highest number of road accidents?</h2>

In [15]:
accident['Accident Date'].value_counts()

Accident Date
30-11-2019    704
31-01-2019    697
13-11-2021    692
13-07-2019    692
14-08-2019    688
             ... 
30-12-2022    171
25-12-2019    157
25-12-2022    145
10/1/2022     123
25-12-2020    118
Name: count, Length: 1461, dtype: int64

<h2>INSIGHT:</h2>

<h3>The result shows that on 30th of November 2019 was recorded the highest number of accidents that makes it a risky period because of the seasonal weather condition.</h3>

<h2>7.Which weather condition has the most accidents?</h2>

In [16]:
accident['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h2>INSIGHT:</h2>

<h3>Based on the result weather conditions does not affect the accident rate because it is high during fine no high winds.</h3>

<h2>8.Which severity is most common in urban areas?</h2>

In [17]:
accident.groupby(['Urban_or_Rural_Area','Accident_Severity']).size()

Urban_or_Rural_Area  Accident_Severity
Rural                Fatal                  5601
                     Serious               37312
                     Slight               196077
Unallocated          Serious                   1
                     Slight                   10
Urban                Fatal                  3060
                     Serious               50902
                     Slight               367701
dtype: int64

<h2>INSIGHT:</h2>

<h3>Based on the data urban areas have the highest number of accidents with 367,701 slight accidents while rural areas has 196,077.This means that urban areas have more accidents overall but usually less severe compared to rural with higher fatality rate.</h3>

<h2>9.How many fatal accidents happen at night vs. daylight?</h2>

In [18]:
accident.groupby(['Light_Conditions','Accident_Severity']).size()

Light_Conditions             Accident_Severity
Darkness - lighting unknown  Fatal                    68
                             Serious                 794
                             Slight                 5622
Darkness - lights lit        Fatal                  1860
                             Serious               19130
                             Slight               108345
Darkness - lights unlit      Fatal                    45
                             Serious                 360
                             Slight                 2138
Darkness - no lighting       Fatal                  1612
                             Serious                7174
                             Slight                28651
Daylight                     Fatal                  5076
                             Serious               60759
                             Slight               419045
dtype: int64

<h2>INSIGHT:</h2>

<h3>The result shows that accidents in darkness without lighting have the highest fatality rate (4.3%), over four times higher than in daylight. While most accidents occur during daylight, they tend to be less severe. </h3>

<h2>10.What is the total accident under fatal severity with one way street? </h2>

In [19]:
accident[accident['Accident_Severity']=='Fatal']['Road_Type'].value_counts()

Road_Type
Single carriageway    6527
Dual carriageway      1815
Roundabout             142
One way street          95
Slip road               49
Name: count, dtype: int64

<h2>INSIGHT:</h2>

<h3>The result shows that the total accident under fatal severity on one way street is 95.</h3>

<h2>11.Which road type has the highest serious accidents?</h2>

In [20]:
accident[accident['Accident_Severity']=='Serious']['Road_Type'].value_counts()

Road_Type
Single carriageway    70059
Dual carriageway      11746
Roundabout             3665
One way street         1655
Slip road               611
Name: count, dtype: int64

<h2>INSIGHT:</h2>

<h3> The road type that has the highest total of serious accidents is on single carriageway with a total of 70,059 accidents.</h3>

<h2>12.How does road surface condition affect accident severity?</h2>

In [21]:
accident.groupby(['Road_Surface_Conditions','Accident_Severity']).size().unstack()

Accident_Severity        Fatal  Serious  Slight
Road_Surface_Conditions                        
Dry                       5788    61638  380395
Flood over 3cm. deep        23      152     842
Frost or ice               193     2007   16317
Snow                        35      565    5290
Wet or damp               2620    23785  160303

<h2>INSIGHT:</h2>

<h3>Accidents on dry roads are the most frequent but tend to be less severe, with a higher proportion of slight injuries.</h3>

<h2>13.Which day of the week has the most accidents?</h2>

In [22]:
accident['Day'] = pd.to_datetime(accident['Accident Date'], dayfirst=True,errors='coerce').dt.day_name()
accident['Day'].value_counts()

Day
Saturday     43164
Wednesday    40037
Friday       39822
Thursday     39641
Tuesday      38714
Sunday       35065
Monday       28564
Name: count, dtype: int64

<h2>INSIGHT:</h2>

<h3>Saturday has the highest number of road accidents, followed by Wednesday and Friday this trend may be due to increased travel during weekends for errands.</h3>

<h2>14.Which vehicle type is most involved in accidents?</h2>

In [23]:
accident['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h2>INSIGHT:</h2>

<h3>The result shows that cars are the most common vehicle prone to accidents. </h3>

<h2>15.Between cars and taxi/private hire car, which are more often involved in accidents?</h2>

In [24]:
accident[accident['Vehicle_Type'].isin(['Car','Taxi/Private hire car'])]['Vehicle_Type'].value_counts()

Vehicle_Type
Car                      497992
Taxi/Private hire car     13294
Name: count, dtype: int64

<h2>INSIGHT:</h2>

<h3>The result shows that cars are most often involved in accidents compared to taxis/private hire cars, likely because they are more common on the road.</h3>

<h2>16.Which district areas report the highest number of fatal,serious and slight accidents?</h2>

In [40]:
accident.groupby(['District Area','Accident_Severity']).size().unstack()

Accident_Severity  Fatal  Serious  Slight
District Area                            
Aberdeen City         12      239    1072
Aberdeenshire         66      463    1401
Adur                   8      101     510
Allerdale             24      143     961
Alnwick                6       33     193
...                  ...      ...     ...
Wychavon              30      193    1138
Wycombe               20      216    1493
Wyre                  15      186    1037
Wyre Forest           22      132     815
York                  21      255    1621

[422 rows x 3 columns]

<h2>INSIGHT:</h2>

<h3>It shows that the District of Aberdeenshire has the higher Fatal and Serious accidents while Wycombe was the highest on slight accidents.</h3>

<h2>17.In which month do most road accidents occur?</h2>

In [26]:
accident['Month']= pd.to_datetime(accident['Accident Date'], dayfirst=True,errors='coerce').dt.month_name()
accident['Month'].value_counts()

Month
November     24240
December     24156
October      23962
July         22939
September    22558
February     22264
June         22196
March        21824
May          21723
August       21106
April        19787
January      18252
Name: count, dtype: int64

<h2>INSIGHT:</h2>

<h3>November has the highest number of accidents,possibly due to colder weather and heavier holiday traffic.</h3>

<h2>18.Is there a correlation between the number of vehicles involved and the number of casualties?</h2>

In [27]:
accident[['Number_of_Casualties','Number_of_Vehicles']].corr()

Number_of_Casualties  Number_of_Vehicles
Number_of_Casualties              1.000000            0.228889
Number_of_Vehicles                0.228889            1.000000

<h2>INSIGHT:</h2>

<h3>There is a weak positive correlation (0.23) between the number of vehicles involved and the number of casualties, indicating that more vehicles slightly increase the likelihood of more casualties.</h3>

<h2>19.What is the average number of vehicles involved under each light condition? </h2>

In [28]:
accident.groupby('Light_Conditions')['Number_of_Vehicles'].mean()

Light_Conditions
Darkness - lighting unknown    1.758482
Darkness - lights lit          1.775498
Darkness - lights unlit        1.752654
Darkness - no lighting         1.599781
Daylight                       1.865385
Name: Number_of_Vehicles, dtype: float64

<h2>INSIGHT:</h2>

<h3>The mean number of vehicles per accident is about 1-2.It shows that daylight are slightly above, while accidents to darkness tend to be at or the ovearall mean. </h3>

<h2>20.How does the average number of casualties per accident vary between urban and rural areas? </h2>

In [37]:
accident.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()

Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287815
Name: Number_of_Casualties, dtype: float64

<h2>INSIGHT:</h2>

<h3>Rural areas have the highest average number of casualties per accident 1.48, compared to urban areas 1.29. This suggests that accidents in rural areas, though possibly less frequent, tend to be more severe, likely due to higher driving speeds and limited lighting.</h3>

<h2>21. Which day of the week have the highest number of fatal accidents?</h2>

In [41]:
accident[accident['Accident_Severity'] == 'Fatal']['Day'].value_counts()

Day
Sunday       577
Monday       561
Saturday     540
Friday       471
Thursday     468
Tuesday      453
Wednesday    434
Name: count, dtype: int64

<h2>INSIGHT:</h2>

<h3>Saturday sees the most fatal accidents, likely due to late-night travel and weekend driving patterns.</h3>

<h2>22.How many accidents occurred during darkness without any street lighting?</h2>

In [42]:
accident['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h2>INSIGHT:</h2>

<h3>Darkness - no lighting" accounts for a smaller share of accidents, but these are often more severe.</h3>

<h2>23.How many serious accidents happened on wet roads?</h2>

In [44]:
accident[(accident['Road_Surface_Conditions'] == 'Wet or damp') & (accident['Accident_Severity'] == 'Serious')].shape[0]

23785

<h2>INSIGHT:</h2>

<h3>There are over 23,785 serious accidents happened when the road was wet or damp.This shows that driving on wet roads is more dangerous, as vehicles are more likely to lose control or take longer to stop, leading to more serious injuries.</h3>

<h2>24.Is there a correlation between number of casualties and number of vehicles by district?</h2>

In [47]:
accident.groupby('District Area')[['Number_of_Vehicles', 'Number_of_Casualties']].mean().corr()

Number_of_Vehicles  Number_of_Casualties
Number_of_Vehicles                1.0000                0.1699
Number_of_Casualties              0.1699                1.0000

<h2>INSIGHTS:</h2>

<h3>Insight 1:
There is a small connection between how many vehicles are involved in an accident and how many people get hurt.
This means that when more vehicles crash, there might be slightly more injuries, but not always.</h3>

<h3>Insight 2:
The number of vehicles in an accident doesn't strongly affect how serious it is. Other things like the speed, road condition, weather, or vehicle type are probably more [possible to caused injuries.</h3>

<h2>25.How does accident severity vary across districts?</h2>

In [56]:
accident.groupby('District Area')['Accident_Severity'].value_counts()

District Area  Accident_Severity
Aberdeen City  Slight               1072
               Serious               239
               Fatal                  12
Aberdeenshire  Slight               1401
               Serious               463
                                    ... 
Wyre Forest    Serious               132
               Fatal                  22
York           Slight               1621
               Serious               255
               Fatal                  21
Name: count, Length: 1266, dtype: int64

<h2>INSIGHTS:</h2>

<h3>Insight 1:
Some districts may have a higher number of slight accidents, indicating more frequent but less severe crashes—likely due to higher traffic density or urban settings.</h3>

<h3>Insight 2:
Other districts might show relatively more serious or fatal accidents, suggesting factors like higher speeds, rural roads, or poorer road conditions that increase accident severity.</h3>